replication/row_event.go

```diff
@@ -946,43 +1014,68 @@ func isBitSet(bitmap []byte, i int) bool {
 	return bitmap[i>>3]&(1<<(uint(i)&7)) > 0
 }
 
-func (e *RowsEvent) decodeRows(data []byte, table *TableMapEvent, bitmap []byte) (int, error) {
-	row := make([]interface{}, e.ColumnCount)
-	skips := make([]int, 0)
+func isBitSetIncr(bitmap []byte, i *int) bool {
+	v := isBitSet(bitmap, *i)
+	*i++
+	return v
+}
+
+func (e *RowsEvent) decodeImage(data []byte, table *TableMapEvent, bitmap []byte, rowImageType EnumRowImageType) (int, error) {
+	// Rows_log_event::print_verbose_one_row()
 
 	pos := 0
 
+	var isPartialJsonUpdate bool
+
+	var partialBitmap []byte
+	if e.eventType == PARTIAL_UPDATE_ROWS_EVENT && rowImageType == EnumRowImageTypeUpdateAI {
+		binlogRowValueOptions, _, n := LengthEncodedInt(data[pos:]) // binlog_row_value_options
+		pos += n
+		isPartialJsonUpdate = EnumBinlogRowValueOptions(binlogRowValueOptions)&EnumBinlogRowValueOptionsPartialJsonUpdates != 0
+		if isPartialJsonUpdate {
+			byteCount := bitmapByteSize(int(e.Table.JsonColumnCount()))
+			partialBitmap = data[pos : pos+byteCount]
+			pos += byteCount
+		}
+	}
+
+	row := make([]interface{}, e.ColumnCount)
+	skips := make([]int, 0)
+
 	// refer: https://github.com/alibaba/canal/blob/c3e38e50e269adafdd38a48c63a1740cde304c67/dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/RowsLogBuffer.java#L63
 	count := 0
 	for i := 0; i < int(e.ColumnCount); i++ {
 		if isBitSet(bitmap, i) {
 			count++
 		}
 	}
-	count = (count + 7) / 8
+	count = bitmapByteSize(count)
 
 	nullBitmap := data[pos : pos+count]
 	pos += count
 
+	isPartialIndex := 0
 	nullbitIndex := 0
 
-	var n int
-	var err error
 	for i := 0; i < int(e.ColumnCount); i++ {
+		isPartial := isPartialJsonUpdate &&
```

Comment:
seems `isPartialIndex` is same as `i`? and we can move creating `isPartial` closer to `e.decodeValue(..., isPartial)`

Reply:
>seems isPartialIndex is same as i?

No, it's an index inside a partial bitmap and we increment it only for partial fields.

>can move creating isPartial closer to e.decodeValue(..., isPartial)

[Partial bitmap checking must prepend to null bitmap checking](https://github.com/mysql/mysql-server/blob/1bfe02bdad6604d54913c62614bde57a055c8332/sql/log_event.cc#L2197-L2203).

I'll update this code for more clearance.